### PR TITLE
[ci/multinode] Add utilities to kill nodes in multi node testing

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Killing nodes enables advanced fault tolerance testing. This PR adds utilities and a test for this functionality in fake multinode docker mode.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/fake_multi_node/test_utils.py [hunk 1]
```diff
@@ -240,3 +245,66 @@ def stop(self):
...
+    def _update_nodes(self):
+        with open(self._nodes_file, "rt") as f:
+            self._nodes = json.load(f)
+
+    def _update_status(self):
+        with open(self._status_file, "rt") as f:
+            self._status = json.load(f)
+
+    def _get_node(self,
+                  node_id: Optional[str] = None,
```
Comment (xwjiang2010) [inline python/ray/autoscaler/_private/fake_multi_node/test_utils.py:258]: Nit: For me `num` and `rand` would be necessary. Not sure if I would use `node_id`.

Comment (krfricke) [reply python/ray/autoscaler/_private/fake_multi_node/test_utils.py:258]: `num` is just a wrapper, but `node_id` can be useful if people fetch node information (e.g. object locations are given with node ids)

File: python/ray/autoscaler/_private/fake_multi_node/test_utils.py [hunk 2]
```diff
@@ -240,3 +245,66 @@ def stop(self):
...
+        If ``node_id`` is given, kill that node.
+
+        If ``num`` is given, construct node_id from this number, and kill
+        that node.
+
+        If ``rand`` is given (as either ``worker`` or ``any``), kill a random
+        node.
+        """
+        node_id = self._get_node(node_id=node_id, num=num, rand=rand)
+        container = self._get_docker_container(node_id=node_id)
```
Comment (xwjiang2010) [inline python/ray/autoscaler/_private/fake_multi_node/test_utils.py:309]: Just curious, does `docker kill` take `node id` at all?

**Review by xwjiang2010 — APPROVED** [review]

File: python/ray/autoscaler/_private/fake_multi_node/test_utils.py (hunk 2 above)
Comment (krfricke) [reply python/ray/autoscaler/_private/fake_multi_node/test_utils.py:309]: Unfortunately no, and there is also no canonical cross-platform mapping to container IDs, so we need to fetch the status here

File: python/ray/autoscaler/_private/fake_multi_node/test_utils.py [hunk 3]
```diff
@@ -240,3 +245,66 @@ def stop(self):
...
+
+    def _get_docker_container(self, node_id: str) -> Optional[str]:
+        self._update_status()
+        node_status = self._status.get(node_id)
+        if not node_status:
+            return None
+
+        return node_status["Name"]
+
+    def kill_node(self,
```
Comment (xwjiang2010) [inline python/ray/autoscaler/_private/fake_multi_node/test_utils.py:294]: Is this supposed to be called from main test thread?

See  a try I did (https://github.com/ray-project/ray/pull/21516/commits/d139750dcbe1f93f758d4a2e458777993fdb167a). The main test thread is occupied by tune.run when we want to kill a docker node. I used another process to kill it. I wonder if I could leverage this API somehow.

Comment (krfricke) [reply python/ray/autoscaler/_private/fake_multi_node/test_utils.py:294]: The way we currently do it in other tests is using callbacks, would that work for you?

**Pull request merged**